### PR TITLE
Add LiBai ViT

### PR DESCRIPTION
**Pull request by rentainhe**
## TODO
本次PR需要完成的内容:
- [x] 给`TransformerLayer`增加`drop_path_prob`参数
- [x] 增加`drop_path` layer
- [x] 增加`PatchEmbedding`
- [x] 使用LiBai内置的Layer搭建ViT模型
- [x] 跑通Loss对齐
- [x] 跑通单测
- [x] 跑通ImageNet训练和测试

## 模型初始化细节
- 这边跟踪了一下DeiT原github里的引用，回溯到timm 0.3.2版本, ViT中的相关初始化
```python
trunc_normal_(self.pos_embed, std=.02)
trunc_normal_(self.cls_token, std=.02)
self.apply(self._init_weights)

def _init_weights(self, m):
    if isinstance(m, nn.Linear):
        trunc_normal_(m.weight, std=.02)
        if isinstance(m, nn.Linear) and m.bias is not None:
            nn.init.constant_(m.bias, 0)
    elif isinstance(m, nn.LayerNorm):
        nn.init.constant_(m.bias, 0)
        nn.init.constant_(m.weight, 1.0)
```
```python
layernorm: eps=1e-6
```

## 权重转换细节
```python
def filter_keys(key, value):
    if "norm1" in key:
        key = key.replace("norm1", "input_layernorm")
    elif "attn.qkv" in key:
        key = key.replace("attn.qkv", "self_attention.query_key_value")
        if "weight" in key:
            value = value.transpose((-1, -2))
    elif "attn.proj" in key:
        key = key.replace("attn.proj", "self_attention.dense")
        if "weight" in key:
            value = value.transpose((-1, -2))
    elif "norm2" in key:
        key = key.replace("norm2", "post_attention_layernorm")
    elif "mlp.fc1" in key:
        key = key.replace("mlp.fc1", "mlp.dense_h_to_4h")
        if "weight" in key:
            value = value.transpose((-1, -2))
    elif "mlp.fc2" in key:
        key = key.replace("mlp.fc2", "mlp.dense_4h_to_h")
        if "weight" in key:
            value = value.transpose((-1, -2))
    elif "head.weight" in key:
            value = value.transpose((-1, -2))
    return key, value

def load_from_torch(model, path="/home/rentianhe/code/OneFlow-Models/libai/vit_tiny_torch_weight.pth"):
    torch_dict = torch.load(path)
    parameters = torch_dict
    new_parameters = dict()
    for key, value in parameters.items():
        if "num_batches_tracked" not in key:
          val = value.detach().cpu().numpy()
          # to global tensor
          key, val = filter_keys(key, val)
          val = flow.tensor(val).to_global(sbp=flow.sbp.broadcast, placement=flow.placement("cuda", {0: range(1)}))
          new_parameters[key] = val
    model.load_state_dict(new_parameters)
    print("successfully load pytorch weight")
    return model
```

### FAQs
- LiBai中的Linear层权重和nn.Linear保存时是相反的，所以在载入的时候需要transpose一下，不知道这个点需不需要修改，有待讨论

### Loss Compare结果
of_torch_diff
of_torch_loss

### CODE REVIEW

File: configs/common/models/vit/vit_tiny_patch16_224.py [hunk 1]
```diff
@@ -11,9 +11,8 @@
     mlp_ratio=4.0,
     depth=12,
     num_heads=3,
-    qkv_bias=True,
     drop_rate=0.0,
     attn_drop_rate=0.0,
```
Comment (L1aoXingyu) [inline configs/common/models/vit/vit_tiny_patch16_224.py:15]: 这两个 drop_rate 默认都是 0 吗

File: libai/models/vision_transformer.py [hunk 2]
```diff
@@ -140,228 +37,127 @@ def __init__(
...
+        self.norm = LayerNorm(embed_dim)
+        self.head = Linear(embed_dim, num_classes)
 
         # Loss func
         self.loss_func = nn.CrossEntropyLoss() if loss_func is None else loss_func
 
-        self.init_weights(weight_init)
+        # weight init
+        trunc_normal_(self.pos_embed, std=0.02)
+        trunc_normal_(self.cls_token, std=0.02)
```
Comment (L1aoXingyu) [inline libai/models/vision_transformer.py:105]: 要不要把这两个初始化移到对应权重定义的后面

File: libai/models/vision_transformer.py [hunk 3]
```diff
@@ -140,228 +37,127 @@ def __init__(
...
+        self.apply(self._init_weights)
+
+    def _init_weights(self, m):
+        if isinstance(m, Linear):
+            trunc_normal_(m.weight, std=0.02)
+            if isinstance(m, Linear) and m.bias is not None:
```
Comment (L1aoXingyu) [inline libai/models/vision_transformer.py:111]: 为什么要重复判断一次 m 是 linear

File: libai/models/vision_transformer.py [hunk 4]
```diff
@@ -140,228 +37,127 @@ def __init__(
...
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
 
     @classmethod
-    def from_config(cls, cfg):
+    def from_config(self, cfg):
```
Comment (L1aoXingyu) [inline libai/models/vision_transformer.py:118]:
```suggestion
    def from_config(cls, cfg):
```

File: libai/models/vision_transformer.py [hunk 5]
```diff
@@ -140,228 +37,127 @@ def __init__(
...
             "embed_dim": cfg.embed_dim,
             "depth": cfg.depth,
             "num_heads": cfg.num_heads,
             "mlp_ratio": cfg.mlp_ratio,
-            "qkv_bias": cfg.qkv_bias,
             "drop_rate": cfg.drop_rate,
             "attn_drop_rate": cfg.attn_drop_rate,
             "drop_path_rate": cfg.drop_path_rate,
+            "num_classes": cfg.num_classes,
+            "loss_func": cfg.loss_func,
```
Comment (L1aoXingyu) [inline libai/models/vision_transformer.py:131]: 这里需要在 config 里面实例化这个 loss func 吗？还是利用 lazycall 的方式

File: libai/models/vision_transformer.py [hunk 6]
```diff
@@ -140,228 +37,127 @@ def __init__(
...
-
     def forward_features(self, x):
-        # position embedding
+        # patch embedding
         x = self.patch_embed(x)
 
         cls_token = self.cls_token.expand(
             x.shape[0], -1, -1
         )  # stole cls_tokens impl from Phil Wang, thanks
         cls_token = cls_token.to_global(sbp=flow.sbp.split(0), placement=cls_token.placement)
```
Comment (L1aoXingyu) [inline libai/models/vision_transformer.py:141]: cls_token 的 sbp 应该和 x 一直，这里只写了 s(0)

File: libai/models/vision_transformer.py [hunk 7]
```diff
@@ -140,228 +37,127 @@ def __init__(
...
-                (cls_token, self.dist_token.expand(x.shape[0], -1, -1), x).to_global(
-                    sbp=flow.sbp.split(0), placement=self.dist_token.placement
-                ),
-                dim=1,
-            )
+        x = flow.cat((cls_token, x), dim=1)
+
+        # position embedding
         pos_embed = self.pos_embed.expand(x.shape[0], -1, -1)
         pos_embed = pos_embed.to_global(sbp=flow.sbp.split(0), placement=pos_embed.placement)
```
Comment (L1aoXingyu) [inline libai/models/vision_transformer.py:146]: 这里也是和上面相同的问题

File: configs/common/models/vit/vit_tiny_patch16_224.py (hunk 1 above)
Comment (rentainhe) [reply configs/common/models/vit/vit_tiny_patch16_224.py:15]: 是的，这两个位置默认是0.0

File: libai/models/vision_transformer.py (hunk 3 above)
Comment (rentainhe) [reply libai/models/vision_transformer.py:111]: 好像把第二行改成`m.bias`也是一样的，我去改一下（不太知道为什么timm里这么做的233

File: libai/models/vision_transformer.py (hunk 4 above)
Comment (rentainhe) [reply libai/models/vision_transformer.py:118]: 好👌

File: libai/models/vision_transformer.py (hunk 5 above)
Comment (rentainhe) [reply libai/models/vision_transformer.py:131]: loss_func可以不加

File: libai/models/vision_transformer.py (hunk 2 above)
Comment (rentainhe) [reply libai/models/vision_transformer.py:105]: 是指`self.apply(self._init_weights)`之后吗，但是好像在前在后没什么区别2333

Comment (L1aoXingyu) [reply libai/models/vision_transformer.py:105]: 是指在 pos_embed 定义之后，就初始化它

Comment (rentainhe) [reply libai/models/vision_transformer.py:105]: emmm，我感觉可能和初始化相关的函数放在一起更加合适
```python
# weight init
trunc_normal_(self.pos_embed, std=0.02)
trunc_normal_(self.cls_token, std=0.02)
self.apply(self._init_weights)
```

**Review by L1aoXingyu — APPROVED** [review]

File: libai/layers/embedding.py [hunk 8]
```diff
@@ -219,3 +219,58 @@ def forward(self, position_ids):
...
+    """2D Image to Patch Embedding
+    Arguments:
+        img_size: size of input image. Defaults to 224.
+        patch_size: embedded patch size. Defaults to 16.
+        in_chans: input channel's size. Defaults to 3.
+        embed_dim: dimension of embedded patch. Default to 768.
+        norm_layer: normalization patch embedding or not. Default to None.
+        flatten: flatten patch embedding or keep the 2-D shape. Default to True.
+        layer_idx: A layer_idx sign which determines the placement. It will be used in pipeline
+        parallelism. Defaults to 0.
```
Comment (CPFLAME) [inline libai/layers/embedding.py:234]: 这里有一些是Default,  有一些是Defaults, 可以统一下

**Review by CPFLAME — APPROVED** [review]

File: libai/layers/embedding.py (hunk 8 above)
Comment (rentainhe) [reply libai/layers/embedding.py:234]: 好，我这边统一一下描述
